tracer/src/Datadog.Trace/Iast/Ranges.cs

```diff
@@ -206,6 +207,45 @@ internal static Range[] CopyWithMark(Range[] ranges, SecureMarks secureMarks)
         return newRanges.ToArray();
     }
 
+    internal static Range[]? GetUnsafeRanges(Range[] ranges, SecureMarks safeMarks, SourceType[]? safeSources)
+    {
+        if (safeMarks == SecureMarks.None && safeSources is null)
+        {
+            return ranges;
+        }
+
+        bool existsSecureRanges = false;
+        for (int x = 0; x < ranges.Length; x++)
+        {
+            var range = ranges[x];
+            if (range.IsMarked(safeMarks) || range.IsSafeSource(safeSources))
+            {
+                existsSecureRanges = true;
+                break;
+            }
+        }
+
+        if (!existsSecureRanges)
+        {
+            // This is made in order to avoid unnecesary allocations (most common situation)
+            return ranges;
+        }
+
+        List<Range> insecureRanges = new List<Range>(ranges.Length);
+        for (int x = 0; x < ranges.Length; x++)
+        {
+            var range = ranges[x];
+            if (range.IsMarked(safeMarks) || range.IsSafeSource(safeSources))
+            {
+                continue;
+            }
+
+            insecureRanges.Add(range);
+        }
+
+        return insecureRanges.ToArray();
```

Comment:
If you want to reduce allocations here, you could
- In the previous `for` loop, don't `break` early. Just count the number of unsafe ranges.
- Instead of using a list, allocate the Range[] with the correct size.

_Might_ actually be easier to follow too, but it's just a nit 🤷‍♂️

Reply:
I was torn between reducing allocations and not checking twice here...